silkrpc/core/estimate_gas_oracle.cpp

```diff
@@ -0,0 +1,131 @@
+/*
+    Copyright 2021 The Silkrpc Authors
+
+    Licensed under the Apache License, Version 2.0 (the "License");
+    you may not use this file except in compliance with the License.
+    You may obtain a copy of the License at
+
+        http://www.apache.org/licenses/LICENSE-2.0
+
+    Unless required by applicable law or agreed to in writing, software
+    distributed under the License is distributed on an "AS IS" BASIS,
+    WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+    See the License for the specific language governing permissions and
+    limitations under the License.
+*/
+
+#include "estimate_gas_oracle.hpp"
+
+#include <algorithm>
+#include <utility>
+
+#include <asio/compose.hpp>
+#include <asio/post.hpp>
+#include <asio/use_awaitable.hpp>
+#include <silkrpc/core/blocks.hpp>
+#include <silkrpc/core/rawdb/chain.hpp>
+
+#include <silkrpc/common/log.hpp>
+
+namespace silkrpc::ego {
+
+asio::awaitable<intx::uint256> EstimateGasOracle::estimate_gas(const Call& call, uint64_t block_number) {
+    SILKRPC_DEBUG << "EstimateGasOracle::estimate_gas called\n";
+
+    std::uint64_t hi;
+    std::uint64_t lo = kTxGas - 1;
+
+    if (call.gas.value_or(0) >= kTxGas) {
+        SILKRPC_DEBUG << "Set HI with gas in args: " << call.gas.value_or(0) << "\n";
+        hi = call.gas.value();
+    } else {
+        // TODO(sixtysixter) worths it to test if returned header is not null?
```

Comment:
As always it depends on the interface contract of `BlockProvider` functor: its current definition returns a `BlockHeader` by value, so it doesn't seem the case it can be null. How are errors supposed to come from `BlockProvider`? If an exception is raised, here the code is fine as it is

Reply:
Right. There are exception. It is fine in this way